File: go/234.palindrome-linked-list.go
/*
 * @lc app=leetcode id=234 lang=golang
 *
 * [234] Palindrome Linked List
 */
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func isPalindrome(head *ListNode) bool {
	return FollowUp(head)
}

// O(n) time complexity and O(1) space complexity
func FollowUp(head *ListNode) bool {
    // 1. get length O(n)
    current := head
    length := 0
    for current != nil {
        length++
        current = current.Next
    }
    if length == 0 || length == 1 {
        return true
    }
    
    // 2. split list O(n)
    var left, right *ListNode
    left = head
    current = left
    for i := 0; i < length / 2 - 1; i++ {
        current = current.Next
    }
    if length % 2 == 1 {
        right = current.Next.Next
    } else {
        right = current.Next
    }
    current.Next = nil
    
    // 3. reverse the second list O(n)
    right = reverse(right)
    
    // 4. compare node by node
    for left != nil && right != nil {
        if left.Val != right.Val {
            return false
        }
        left = left.Next
        right = right.Next
    }
    return true
}

func reverse(head *ListNode) *ListNode {
    // empty list
    if head == nil {
        return nil
    }
    
    current := head.Next
    previous := head
   
    // single node
    if current == nil {
        return head
    }
   
    previous.Next = nil
    for current != nil {
        current, previous, current.Next = current.Next, current, previous
    }
    
    return previous
}

func Origin(head *ListNode) bool {
	var list []int
	for head != nil {
		list = append(list, head.Val)
		head = head.Next
	}

	for i, j := 0, len(list)-1; i < j; {
		if list[i] != list[j] {
			return false
		}
		i++
		j--
	}

	return true
}

File: csharp/234.palindrome-linked-list.cs
/*
 * @lc app=leetcode id=234 lang=csharp
 *
 * [234] Palindrome Linked List
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int val=0, ListNode next=null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */
public class Solution 
{
    // O(n)
    public bool IsPalindrome(ListNode head) 
    {
        // get length
        int length = 0;
        ListNode cur = head;
        while (cur != null)
        {
            cur = cur.next;
            length++;
        }

        // single element list is a palindrome, length guaranteed no less than 1
        if (length == 1)
        {
            return true;
        }

        // reverse first half of the list
        var (left, right) = this.ReverseLinkedList(head, length / 2);
        // ignore the center element
        if (length % 2 != 0)
        {
            right = right.next;
        }

        // compare two (half) list
        while (left != null) // right != null
        {
            if (left.val != right.val)
            {
                return false;
            }
            left = left.next;
            right = right.next;
        }

        return true;
    }

    // Reverse first 'count' element in a list
    // (head of the reversed list, head of the rest list)
    public (ListNode, ListNode) ReverseLinkedList(ListNode head, int count)
    {
        int i = 0;
        
        ListNode prev, cur, next;
        next = null;
        prev = null;
        cur = head;
        while (i < count)
        {
            next = cur.next;
            cur.next = prev;
            prev = cur;
            cur = next;
            i++;
        }

        return (prev, cur);
    }
}
// @lc code=end

